# Jogadores - América do Sul

In [1]:
# importando bibliotecas
import requests
import pandas as pd
import numpy as np
import dotenv
import os

In [2]:
# carregando arquivo .env com as chaves da API
dotenv.load_dotenv("headers.env")

True

In [3]:
# definindo as chaves da API em variáveis
x=os.getenv("x-rapidapi-key")
y=os.getenv("x-rapidapi-host")
headers = {"x-rapidapi-key": x, "x-rapidapi-host": y}

## API - Criando Dataframe dos jogadores

- **Países**

In [4]:
# criando função para os países
def paises():
    
    """
    Retorna um Dataframe com todos os id dos campeonatos de 2020 dos países disponíveis da América do Sul
    """

    paises=["argentina", "brazil", "uruguay", "paraguay", "venezuela", "peru", "bolivia", "colombia", "ecuador", "chile"]
    df=pd.DataFrame()
    for i in paises:
        url_ligas=f"https://api-football-v1.p.rapidapi.com/v2/leagues/country/{i}/2020"
        response=requests.get(url_ligas, headers=headers)
        dicionario=pd.json_normalize(response.json())["api.leagues"][0]
        df_temp=pd.DataFrame(dicionario)
        df=df.append(df_temp)
    return df

In [5]:
# aumentando o número de linhas na visualização do Dataframe
pd.options.display.max_rows=99

In [6]:
# verificando o id das ligas em interesse
paises()

league_id                         name    type    country country_code  \
0        1291               Copa Argentina     Cup  Argentina           AR   
1        1400         Copa de la Superliga     Cup  Argentina           AR   
2        3023             Primera Division  League  Argentina           AR   
3        3052           Primera B Nacional  League  Argentina           AR   
4        3053                    Primera C  League  Argentina           AR   
5        3054      Primera B Metropolitana  League  Argentina           AR   
6        3055                    Primera D  League  Argentina           AR   
7        3059             Torneo Federal A  League  Argentina           AR   
0        1246                     Alagoano     Cup     Brazil           BR   
1        1317             Brasileiro Women  League     Brazil           BR   
2        1333               Copa Do Brasil     Cup     Brazil           BR   
3        1345                Paulista - A1  League     Brazil           BR   
4        1346                Paulista - A2  League     Brazil           BR   
5        1347                   Gaúcho - 1  League     Brazil           BR   
6        1348                   Gaúcho - 2  League     Brazil           BR   
7        1396                      Serie A  League     Brazil           BR   
8        1397                      Serie B  League     Brazil           BR   
9        1472                      Serie C  League     Brazil           BR   
10       1476                      Serie D  League     Brazil           BR   
11       1529                      Acreano  League     Brazil           BR   
12       1536                    Amapaense  League     Brazil           BR   
13       1541                   Amazonense  League     Brazil           BR   
14       2247                   Baiano - 1  League     Brazil           BR   
15       2248                    Paraibano  League     Brazil           BR   
16       2249              Catarinense - 1  League     Brazil           BR   
17       2250                Paulista - A3  League     Brazil           BR   
18       2251               Paranaense - 1  League     Brazil           BR   
19       2252                   Roraimense  League     Brazil           BR   
20       2253                   Maranhense  League     Brazil           BR   
21       2254                 Cearense - 1  League     Brazil           BR   
22       2255                  Brasiliense  League     Brazil           BR   
23       2256                     Capixaba  League     Brazil           BR   
24       2257             Copa do Nordeste     Cup     Brazil           BR   
25       2259               Paranaense - 2  League     Brazil           BR   
26       2260                  Rondoniense  League     Brazil           BR   
27       2261                     Potiguar  League     Brazil           BR   
28       2262           Copa do Brasil U20     Cup     Brazil           BR   
29       2263          São Paulo Youth Cup     Cup     Brazil           BR   
30       2264                  Mineiro - 2  League     Brazil           BR   
31       2266                    Piauiense  League     Brazil           BR   
32       2267             Pernambucano - 1  League     Brazil           BR   
33       2268            Sul-Matogrossense  League     Brazil           BR   
34       2269                  Carioca - 1  League     Brazil           BR   
35       2271                    Sergipano  League     Brazil           BR   
36       2272                     Paraense  League     Brazil           BR   
37       2273                   Goiano - 1  League     Brazil           BR   
38       2274                  Mineiro - 1  League     Brazil           BR   
39       2275                Matogrossense  League     Brazil           BR   
40       2276                 Tocantinense  League     Brazil           BR   
41       2277          Supercopa do Brasil     Cup     Brazil           BR   
42       2651             

In [7]:
# lista com id das ligas em interesse
id_ligas=[3023, 1396, 1397, 1375, 1356, 1350, 1341, 1340, 1326, 1262, 1342]

- **Times**

In [8]:
# criando função para os times
def times():
    
    """
    Retorna um Dataframe com os times das ligas em interesse
    """
    
    df=pd.DataFrame()
    for i in id_ligas:
        url_times=f"https://api-football-v1.p.rapidapi.com/v2/teams/league/{i}"
        response=requests.get(url_times, headers=headers)
        dicionario=pd.json_normalize(response.json())["api.teams"][0]
        df_temp=pd.DataFrame(dicionario)
        df=df.append(df_temp)
    return df

In [9]:
# aumentando o número de linhas na visualização do Dataframe
pd.options.display.max_rows=400

In [10]:
# verificando os times obtidos
id_times=times().reset_index().loc[:, ["team_id", "name"]]
id_times

team_id                         name
0        436                  Racing Club
1        455             Atletico Tucuman
2        441               Union Santa Fe
3        459              Arsenal Sarandi
4       1065  Central Cordoba de Santiago
5        453                Independiente
6        442           Defensa Y Justicia
7        448               Colon Santa Fe
8        435                  River Plate
9        449                     Banfield
10       437              Rosario Central
11       439                   Godoy Cruz
12       434                Gimnasia L.P.
13       444                    Patronato
14       438              Velez Sarsfield
15       445                      Huracan
16       456             Talleres Cordoba
17       457             Newells Old Boys
18       446                        Lanus
19       451                 Boca Juniors
20       458               Argentinos JRS
21       460                  San Lorenzo
22       463                     Aldosivi
23       450             Estudiantes L.P.
24       118                        Bahia
25       119                Internacional
26       120                     Botafogo
27       121                    Palmeiras
28       123                 Sport Recife
29       124                   Fluminense
30       126                    Sao Paulo
31       127                     Flamengo
32       128                       Santos
33       129                        Ceara
34       130                       Gremio
35       131                  Corinthians
36       133                Vasco DA Gama
37       134          Atletico Paranaense
38       144          Atletico Goianiense
39       147                     Coritiba
40       151                        Goias
41       154                 Fortaleza EC
42       794                   Bragantino
43      1062                  Atletico-MG
44       122                       Parana
45       125              America Mineiro
46       132               Chapecoense-sc
47       135                     Cruzeiro
48       136                      Vitoria
49       137                  Figueirense
50       138             Guarani Campinas
51       139                  Ponte Preta
52       141            Brasil DE Pelotas
53       143                        Oeste
54       145                         Avai
55       146                          CRB
56       150                          CSA
57       152                    Juventude
58       155               Sampaio Correa
59       755               Nautico Recife
60      1193                       Cuiaba
61      1223                  Operario-PR
62      2618                  Botafogo SP
63      7772                    Confiança
64      2365              Atletico Torque
65      2363                     Progreso
66      2351               CA River Plate
67      2357                        Fenix
68      2348                      Penarol
69      2362                        Cerro
70      2360                    Wanderers
71      2350            Defensor Sporting
72      2369                  Cerro Largo
73      2352                      Danubio
74      2371                    Rentistas
75      2356                Club Nacional
76      2370          Deportivo Maldonado
77      2361                 Boston River
78      2358         Liverpool Montevideo
79      2355                Plaza Colonia
80      1174                 Club Guarani
81      1175            Nacional Asuncion
82      1176                Cerro Porteno
83      1177               SOL DE America
84      1178                 General Diaz
85      1179            Libertad Asuncion
86      1182                      Olimpia
87      1183             Sportivo Luqueno
88      1184             Club River Plate
89      2125         Club Sp. San Lorenzo
90      2138                  Guairena FC
91      3832                12 de Octubre
92      2823          Llaneros de Guanare
93      2813             Real Esppor Club
94      2812   

- **Jogadores**

In [11]:
# criando função para os jogadores
def jogadores():
    
    """
    Retorna um Dataframe com os jogadores dos times em interesse
    """
    
    df=pd.DataFrame()
    for i in id_times["team_id"].values:
        url_jogadores=f"https://api-football-v1.p.rapidapi.com/v2/players/team/{i}/2020"
        response=requests.get(url_jogadores, headers=headers)
        dicionario=pd.json_normalize(response.json())["api.players"][0]
        df_temp=pd.DataFrame(dicionario)
        df=df.append(df_temp)
    return df

In [12]:
# definindo o Dataframe dos jogadores em uma variável
jogadores=jogadores()

In [13]:
# filtrando apenas as ligas em interesse respectivas aos jogadores obtidos
jogadores=jogadores.loc[jogadores["league_id"].isin(id_ligas), :].reset_index()

In [14]:
# aumentando o número de colunas na visualização do Dataframe
pd.options.display.max_columns=99

In [15]:
# verificando o Dataframe dos jogadores
jogadores

index  player_id                  player_name          firstname  \
0         1     197252  Jose Roberto Ramos da Silva       Jose Roberto   
1         2     238471                    Christian          Christian   
2        10        374                   João Pedro         João Pedro   
3        17      10033                       Flavio             Flavio   
4        21      10017    Ignacio da Silva Oliveira            Ignacio   
...     ...        ...                          ...                ...   
5376     35     201908                   A. Aravena  Alexander Ernesto   
5377     36      11377                  D. Valencia       Diego Martín   
5378     37      11354                  S. Magnasco            Stefano   
5379     38      11368                   C. Pinares      César Ignacio   
5380     39      11365                  B. Kuscevic           Benjamín   

                 lastname number    position  age  birth_date  \
0          Ramos da Silva   None  Midfielder   19  20/06/2001   
1          de Souza André   None    Attacker   19  17/03/2001   
2     Maturano dos Santos   None    Defender   25  15/11/1996   
3       Medeiros da Silva   None  Midfielder   25  10/02/1996   
4       da Silva Oliveira   None    Defender   25  01/12/1996   
...                   ...    ...         ...  ...         ...   
5376       Aravena Guzmán   None    Attacker   19  06/09/2002   
5377     Valencia Morello   None    Attacker   21  14/01/2000   
5378     Magnasco Galindo   None    Defender   29  28/09/1992   
5379       Pinares Tamayo   None  Midfielder   30  23/05/1991   
5380   Kuscevic Jaramillo   None    Defender   25  02/05/1996   

               birth_place birth_country nationality  height weight injured  \
0                     None        Brazil      Brazil  173 cm  60 kg    None   
1                     None        Brazil      Brazil    None   None    None   
2     Presidente Bernardes        Brazil      Brazil  179 cm  71 kg    None   
3                   Santos        Brazil      Brazil  178 cm  73 kg    None   
4                     Assu        Brazil      Brazil  184 cm  75 kg    None   
...                    ...           ...         ...     ...    ...     ...   
5376                  None         Chile       Chile  174 cm  70 kg    None   
5377                  None         Chile       Chile  184 cm  75 kg    None   
5378              Santiago         Chile       Chile  175 cm   None    None   
5379     Santiago de Chile         Chile       Chile  176 cm  75 kg    None   
5380                  None         Chile       Chile  185 cm  74 kg    None   

        rating  team_id    team_name  league_id            league season  \
0         None      118        Bahia     1396.0           Serie A   2020   
1         None      118        Bahia     1396.0           Serie A   2020   
2     6.700000      118        Bahia     1396.0           Serie A   2020   
3     7.300000      118        Bahia     1396.0           Serie A   2020   
4         None      118        Bahia     1396.0           Serie A   2020   
...        ...      ...          ...        ...               ...    ...   
5376  6.300000     2994  U. Catolica     1342.0  Primera Division   2020   
5377  6.730769     2994  U. Catolica     1342.0  Primera Division   2020   
5378  7.125000     2994  U. Catolica     1342.0  Primera Division   2020   
5379  7.077777     2994  U. Catolica     1342.0  Primera Division   2020   
5380  6.680000     2994  U. Catolica     1342.0  Primera Division   2020   

      captain                    shots  \
0           0    {'total': 0, 'on': 0}   
1           0    {'total': 0, 'on': 0}   
2           0    {'total': 0, 'on': 0}   
3           0    {'total': 1, 'on': 1}   
4           0    {'total': 0, 'on': 0}   
...       ...                      ...   
5376        0    {'total': 0, 'on': 0}   
5377        0  {'total': 14, 'on': 11}   
5378        0    {'total': 0, 'on': 0}   
5379        0    {'total': 9, 'on': 4}   
5380        

### Limpando e manipulando o Dataframe obtido

In [16]:
# filtrando as colunas em interesse e redefinindo os nomes
jogadores=jogadores.loc[:, ["player_id", "player_name", "position", "age", "birth_date", "nationality", "team_name", "league_id", "league", "shots", "goals", "passes", "tackles", "dribbles", "fouls", "games"]]
jogadores.columns=["id", "nome", "posição", "idade", "data_nascimento", "nacionalidade", "time", "id_liga", "liga", "chutes", "gols1", "passes", "desarmes1", "dribles", "faltas", "jogos"]

In [17]:
# verificando o Dataframe
jogadores

id                         nome     posição  idade data_nascimento  \
0     197252  Jose Roberto Ramos da Silva  Midfielder     19      20/06/2001   
1     238471                    Christian    Attacker     19      17/03/2001   
2        374                   João Pedro    Defender     25      15/11/1996   
3      10033                       Flavio  Midfielder     25      10/02/1996   
4      10017    Ignacio da Silva Oliveira    Defender     25      01/12/1996   
...      ...                          ...         ...    ...             ...   
5376  201908                   A. Aravena    Attacker     19      06/09/2002   
5377   11377                  D. Valencia    Attacker     21      14/01/2000   
5378   11354                  S. Magnasco    Defender     29      28/09/1992   
5379   11368                   C. Pinares  Midfielder     30      23/05/1991   
5380   11365                  B. Kuscevic    Defender     25      02/05/1996   

     nacionalidade         time  id_liga              liga  \
0           Brazil        Bahia   1396.0           Serie A   
1           Brazil        Bahia   1396.0           Serie A   
2           Brazil        Bahia   1396.0           Serie A   
3           Brazil        Bahia   1396.0           Serie A   
4           Brazil        Bahia   1396.0           Serie A   
...            ...          ...      ...               ...   
5376         Chile  U. Catolica   1342.0  Primera Division   
5377         Chile  U. Catolica   1342.0  Primera Division   
5378         Chile  U. Catolica   1342.0  Primera Division   
5379         Chile  U. Catolica   1342.0  Primera Division   
5380         Chile  U. Catolica   1342.0  Primera Division   

                       chutes  \
0       {'total': 0, 'on': 0}   
1       {'total': 0, 'on': 0}   
2       {'total': 0, 'on': 0}   
3       {'total': 1, 'on': 1}   
4       {'total': 0, 'on': 0}   
...                       ...   
5376    {'total': 0, 'on': 0}   
5377  {'total': 14, 'on': 11}   
5378    {'total': 0, 'on': 0}   
5379    {'total': 9, 'on': 4}   
5380    {'total': 3, 'on': 2}   

                                                  gols1  \
0     {'total': 0, 'conceded': 0, 'assists': 0, 'sav...   
1     {'total': 0, 'conceded': 0, 'assists': 0, 'sav...   
2     {'total': 0, 'conceded': 0, 'assists': 0, 'sav...   
3     {'total': 0, 'conceded': 0, 'assists': 0, 'sav...   
4     {'total': 0, 'conceded': 0, 'assists': 0, 'sav...   
...                                                 ...   
5376  {'total': 0, 'conceded': 0, 'assists': 0, 'sav...   
5377  {'total': 4, 'conceded': 0, 'assists': 1, 'sav...   
5378  {'total': 0, 'conceded': 0, 'assists': 1, 'sav...   
5379  {'total': 4, 'conceded': 0, 'assists': 1, 'sav...   
5380  {'total': 1, 'conceded': 0, 'assists': 0, 'sav...   

                                         passes  \
0         {'total': 0, 'key': 0, 'accuracy': 0}   
1         {'total': 0, 'key': 0, 'accuracy': 0}   
2       {'total': 19, 'key': 0, 'accuracy': 67}   
3       {'total': 39, 'key': 1, 'accuracy': 83}   
4         {'total': 0, 'key': 0, 'accuracy': 0}   
...                                         ...   
5376    {'total': 13, 'key': 0, 'accuracy': 12}   
5377   {'total': 147, 'key': 8, 'accuracy': 21}   
5378   {'total': 331, 'key': 2, 'accuracy': 35}   
5379  {'total': 239, 'key': 13, 'accuracy': 41}   
5380   {'total': 229, 'key': 0, 'accuracy': 86}   

                                            desarmes1  \
0       {'total': 0, 'blocks': 0, 'interceptions': 0}   
1       {'total': 0, 'blocks': 0, 'interceptions': 0}   
2       {'total': 0, 'blocks': 1, 'interceptions': 3}   
3       {'total': 9, 'blocks': 0, 'interceptions': 0}   
4       {'total': 0, 'blocks': 0, 'interceptions': 0}   
...                                               ...   
5376    {'total': 0, 'blocks': 0, 'interceptions': 0}   
5377    {'total': 7, 'blocks': 0, 'interceptions': 2}   
5378  {'total': 14, 'blocks': 0, 'interceptions': 20}   
5379    {'total

In [18]:
# lidando com as colunas que possuem um dicionário em cada linha
a=jogadores.merge(pd.json_normalize(jogadores["chutes"]), on=jogadores.index).drop("key_0", 1)
a.rename(columns={"total": "total_chutes", "on": "chutes_no_gol"}, inplace=True)
b=a.merge(pd.json_normalize(a["gols1"]), on=a.index).drop("key_0", axis=1)
b.rename(columns={"total": "gols", "assists": "assistências", "conceded": "gols_sofridos", "saves": "defesas"}, inplace=True)
c=b.merge(pd.json_normalize(b["passes"]), on=b.index).drop(["key_0"], axis=1)
c.rename(columns={"total": "total_passes", "key": "passes_chaves", "accuracy": "%passes_certos"}, inplace=True)
d=c.merge(pd.json_normalize(c["desarmes1"]), on=c.index).drop(["key_0", "blocks", "interceptions"], axis=1)
d.rename(columns={"total": "desarmes"}, inplace=True)
e=d.merge(pd.json_normalize(d["dribles"]), on=d.index).drop(["key_0"], axis=1)
e.rename(columns={"attempts": "dribles_tentados", "success": "dribles_realizados"}, inplace=True)
f=e.merge(pd.json_normalize(e["faltas"]), on=e.index).drop(["key_0"], axis=1)
f.rename(columns={"drawn": "faltas_sofridas", "committed": "faltas_cometidas"}, inplace=True)
jogadores_ok=f.merge(pd.json_normalize(f["jogos"]), on=f.index).drop(["key_0"], axis=1)
jogadores_ok.rename(columns={"appearences": "total_jogos", "minutes_played": "minutos_jogados", "lineups": "jogos_escalados"}, inplace=True)
jogadores_ok.drop(jogadores_ok.columns[9:16], axis=1, inplace=True)

In [19]:
# verificando o Dataframe
jogadores_ok

id                         nome     posição  idade data_nascimento  \
0     197252  Jose Roberto Ramos da Silva  Midfielder     19      20/06/2001   
1     238471                    Christian    Attacker     19      17/03/2001   
2        374                   João Pedro    Defender     25      15/11/1996   
3      10033                       Flavio  Midfielder     25      10/02/1996   
4      10017    Ignacio da Silva Oliveira    Defender     25      01/12/1996   
...      ...                          ...         ...    ...             ...   
5376  201908                   A. Aravena    Attacker     19      06/09/2002   
5377   11377                  D. Valencia    Attacker     21      14/01/2000   
5378   11354                  S. Magnasco    Defender     29      28/09/1992   
5379   11368                   C. Pinares  Midfielder     30      23/05/1991   
5380   11365                  B. Kuscevic    Defender     25      02/05/1996   

     nacionalidade         time  id_liga              liga  total_chutes  \
0           Brazil        Bahia   1396.0           Serie A             0   
1           Brazil        Bahia   1396.0           Serie A             0   
2           Brazil        Bahia   1396.0           Serie A             0   
3           Brazil        Bahia   1396.0           Serie A             1   
4           Brazil        Bahia   1396.0           Serie A             0   
...            ...          ...      ...               ...           ...   
5376         Chile  U. Catolica   1342.0  Primera Division             0   
5377         Chile  U. Catolica   1342.0  Primera Division            14   
5378         Chile  U. Catolica   1342.0  Primera Division             0   
5379         Chile  U. Catolica   1342.0  Primera Division             9   
5380         Chile  U. Catolica   1342.0  Primera Division             3   

      chutes_no_gol  gols  gols_sofridos  assistências  defesas  total_passes  \
0                 0     0              0             0        0             0   
1                 0     0              0             0        0             0   
2                 0     0              0             0        0            19   
3                 1     0              0             0        0            39   
4                 0     0              0             0        0             0   
...             ...   ...            ...           ...      ...           ...   
5376              0     0              0             0        0            13   
5377             11     4              0             1        0           147   
5378              0     0              0             1        0           331   
5379              4     4              0             1        0           239   
5380              2     1              0             0        0           229   

      passes_chaves  %passes_certos  desarmes  dribles_tentados  \
0                 0               0         0                 0   
1                 0               0         0                 0   
2                 0              67         0                 3   
3                 1              83         9                 1   
4                 0               0         0                 0   
...             ...             ...       ...               ...   
5376              0              12         0                 1   
5377              8              21         7                 9   
5378              2              35        14                 7   
5379             13              41         7                15   
5380              0              86         3                 1   

      dribles_realizados  faltas_sofridas  faltas_cometidas  total_jogos  \
0                      0                0                 0            0   
1                      0                0                 0            0   
2                      2                0                 2            1   
3                      1                1             

In [20]:
# verificando a quantidade de jogadores duplicados
jogadores_ok.duplicated(subset="id").sum()

244

In [21]:
# verificando os jogadores duplicados
jogadores_ok.loc[jogadores_ok.duplicated(subset="id", keep=False), :].sort_values(by="id")

id                           nome     posição  idade  \
686     1065              Natanael Pimienta    Defender     31   
89      1065              Natanael Pimienta    Defender     31   
3952    2848                       J. Ortiz    Attacker     29   
4602    2848                       J. Ortiz    Attacker     29   
4793    5683                     F. Anselmo    Attacker     27   
...      ...                            ...         ...    ...   
1108  216550           Rone Carlos de Paula  Midfielder     26   
1955  238471                      Christian    Attacker     19   
1     238471                      Christian    Attacker     19   
3334  276448  Carlos Alberto Ruíz Gutiérrez    Attacker     19   
3569  276448  Carlos Alberto Ruíz Gutiérrez    Attacker     19   

     data_nascimento nacionalidade                 time  id_liga  \
686       25/12/1990        Brazil  Atletico Goianiense   1396.0   
89        25/12/1990        Brazil        Internacional   1396.0   
3952      20/06/1992    Costa Rica      Deportes Tolima   1326.0   
4602      20/06/1992    Costa Rica          Millonarios   1326.0   
4793      17/04/1994     Argentina            Palestino   1342.0   
...              ...           ...                  ...      ...   
1108      15/08/1995        Brazil       Chapecoense-sc   1397.0   
1955      17/03/2001        Brazil            Confiança   1397.0   
1         17/03/2001        Brazil                Bahia   1396.0   
3334      25/01/2002          Peru                  UTC   1341.0   
3569      25/01/2002          Peru     Sporting Cristal   1341.0   

                  liga  total_chutes  chutes_no_gol  gols  gols_sofridos  \
686            Serie A             5              2     0              0   
89             Serie A             5              2     0              0   
3952         Primera A            19              8     0              0   
4602         Primera A            19              8     1              0   
4793  Primera Division             2              0     0              0   
...                ...           ...            ...   ...            ...   
1108           Serie B             1              0     0              0   
1955           Serie B             0              0     0              0   
1              Serie A             0              0     0              0   
3334  Primera Division             1              0     0              0   
3569  Primera Division             1              0     0              0   

      assistências  defesas  total_passes  passes_chaves  %passes_certos  \
686              0        0           171              3              12   
89               0        0           171              3              12   
3952             0        0            72              7              76   
4602             0        0            72              7              76   
4793             0        0             1              0               0   
...            ...      ...           ...            ...             ...   
1108             0        0            10              0              54   
1955             0        0             0              0               0   
1                0        0             0              0               0   
3334             0        0             2              0               1   
3569             0        0             2              0               1   

      desarmes  dribles_tentados  dribles_realizados  faltas_sofridas  \
686         11                14                   7                8   
89          11                14                   7                8   
3952         2                 3                   1                5   
4602         2                 3                   1                5   
4793         0                 0                   0                1   
...        ...               ...                 ...              ...   
1108         0                 2                   1           

In [22]:
# verificando o id das ligas encontradas/disponíveis em 2020
jogadores_ok.id_liga.unique()

array([1396., 1397., 1375., 1350., 1341., 1326., 1342.])

In [23]:
# verificando o nome das ligas encontradas/disponíveis em 2020
jogadores_ok.liga.unique()

array(['Serie A', 'Serie B', 'Primera Division - Apertura',
       'Primera Division', 'Primera A'], dtype=object)

In [24]:
# redefinindo o nome das ligas
jogadores_ok.loc[jogadores_ok["id_liga"]==1396, "liga"]="Brasileiro Série A"
jogadores_ok.loc[jogadores_ok["id_liga"]==1397, "liga"]="Brasileiro Série B"
jogadores_ok.loc[jogadores_ok["id_liga"]==1375, "liga"]="Liga Uruguai"
jogadores_ok.loc[jogadores_ok["id_liga"]==1350, "liga"]="Liga Venezuela"
jogadores_ok.loc[jogadores_ok["id_liga"]==1341, "liga"]="Liga Peru"
jogadores_ok.loc[jogadores_ok["id_liga"]==1326, "liga"]="Liga Colômbia"
jogadores_ok.loc[jogadores_ok["id_liga"]==1342, "liga"]="Liga Chile"

In [25]:
# verificando o Dataframe
jogadores_ok

id                         nome     posição  idade data_nascimento  \
0     197252  Jose Roberto Ramos da Silva  Midfielder     19      20/06/2001   
1     238471                    Christian    Attacker     19      17/03/2001   
2        374                   João Pedro    Defender     25      15/11/1996   
3      10033                       Flavio  Midfielder     25      10/02/1996   
4      10017    Ignacio da Silva Oliveira    Defender     25      01/12/1996   
...      ...                          ...         ...    ...             ...   
5376  201908                   A. Aravena    Attacker     19      06/09/2002   
5377   11377                  D. Valencia    Attacker     21      14/01/2000   
5378   11354                  S. Magnasco    Defender     29      28/09/1992   
5379   11368                   C. Pinares  Midfielder     30      23/05/1991   
5380   11365                  B. Kuscevic    Defender     25      02/05/1996   

     nacionalidade         time  id_liga                liga  total_chutes  \
0           Brazil        Bahia   1396.0  Brasileiro Série A             0   
1           Brazil        Bahia   1396.0  Brasileiro Série A             0   
2           Brazil        Bahia   1396.0  Brasileiro Série A             0   
3           Brazil        Bahia   1396.0  Brasileiro Série A             1   
4           Brazil        Bahia   1396.0  Brasileiro Série A             0   
...            ...          ...      ...                 ...           ...   
5376         Chile  U. Catolica   1342.0          Liga Chile             0   
5377         Chile  U. Catolica   1342.0          Liga Chile            14   
5378         Chile  U. Catolica   1342.0          Liga Chile             0   
5379         Chile  U. Catolica   1342.0          Liga Chile             9   
5380         Chile  U. Catolica   1342.0          Liga Chile             3   

      chutes_no_gol  gols  gols_sofridos  assistências  defesas  total_passes  \
0                 0     0              0             0        0             0   
1                 0     0              0             0        0             0   
2                 0     0              0             0        0            19   
3                 1     0              0             0        0            39   
4                 0     0              0             0        0             0   
...             ...   ...            ...           ...      ...           ...   
5376              0     0              0             0        0            13   
5377             11     4              0             1        0           147   
5378              0     0              0             1        0           331   
5379              4     4              0             1        0           239   
5380              2     1              0             0        0           229   

      passes_chaves  %passes_certos  desarmes  dribles_tentados  \
0                 0               0         0                 0   
1                 0               0         0                 0   
2                 0              67         0                 3   
3                 1              83         9                 1   
4                 0               0         0                 0   
...             ...             ...       ...               ...   
5376              0              12         0                 1   
5377              8              21         7                 9   
5378              2              35        14                 7   
5379             13              41         7                15   
5380              0              86         3                 1   

      dribles_realizados  faltas_sofridas  faltas_cometidas  total_jogos  \
0                      0                0                 0            0   
1                      0                0                 0            0   
2                      2                0                 2            1   
3                      1      

In [26]:
#jogadores_ok.to_csv("jogadores.csv", index=False)

In [27]:
#jogadores_ok=pd.read_csv("./jogadores.csv")

In [28]:
#jogadores_ok

In [29]:
# filtrando jogadores que jogaram no mínimo 360 minutos
jogadores_ok=jogadores_ok.loc[(jogadores_ok["minutos_jogados"]>=360), :]

In [30]:
# criando colunas que consideram os scouts com base nos minutos jogados
jogadores_ok["chutes/min"]=jogadores_ok["total_chutes"]/jogadores_ok["minutos_jogados"]
jogadores_ok["chutes_no_gol/min"]=jogadores_ok["chutes_no_gol"]/jogadores_ok["minutos_jogados"]
jogadores_ok["gols/min"]=jogadores_ok["gols"]/jogadores_ok["minutos_jogados"]
jogadores_ok["sofridos/min"]=jogadores_ok["gols_sofridos"]/jogadores_ok["minutos_jogados"]
jogadores_ok["assistências/min"]=jogadores_ok["assistências"]/jogadores_ok["minutos_jogados"]
jogadores_ok["defesas/min"]=jogadores_ok["defesas"]/jogadores_ok["minutos_jogados"]
jogadores_ok["passes/min"]=jogadores_ok["total_passes"]/jogadores_ok["minutos_jogados"]
jogadores_ok["passes_chaves/min"]=jogadores_ok["passes_chaves"]/jogadores_ok["minutos_jogados"]
jogadores_ok["desarmes/min"]=jogadores_ok["desarmes"]/jogadores_ok["minutos_jogados"]
jogadores_ok["dribles_tentados/min"]=jogadores_ok["dribles_tentados"]/jogadores_ok["minutos_jogados"]
jogadores_ok["dribles_realizados/min"]=jogadores_ok["dribles_realizados"]/jogadores_ok["minutos_jogados"]
jogadores_ok["faltas_sofridas/min"]=jogadores_ok["faltas_sofridas"]/jogadores_ok["minutos_jogados"]
jogadores_ok["faltas_cometidas/min"]=jogadores_ok["faltas_cometidas"]/jogadores_ok["minutos_jogados"]

<ipython-input-30-cdd8d35fcbb2>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  jogadores_ok["chutes/min"]=jogadores_ok["total_chutes"]/jogadores_ok["minutos_jogados"]
<ipython-input-30-cdd8d35fcbb2>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  jogadores_ok["chutes_no_gol/min"]=jogadores_ok["chutes_no_gol"]/jogadores_ok["minutos_jogados"]
<ipython-input-30-cdd8d35fcbb2>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer]

In [31]:
# criando colunas numéricas para as colunas sobre liga e nacionalidade
# jogadores_ok=jogadores_ok.merge(pd.get_dummies(jogadores_ok["id_liga"]), on=jogadores_ok.index).drop("key_0", axis=1)
# jogadores_ok=jogadores_ok.merge(pd.get_dummies(jogadores_ok["nacionalidade"]), on=jogadores_ok.index).drop("key_0", axis=1)

## Criando a similaridade dos jogadores

In [32]:
# importando bibliotecas de distâncias
from scipy.spatial.distance import pdist, squareform

In [33]:
# criando função de similaridade
def criar_similaridade(dataframe, metric="euclidean"):
    """
    Retorna a matriz similar de um Dataframe
    """
    distancia_vetor=pdist(dataframe, metric)
    distancia_matriz=squareform(distancia_vetor)
    similaridade=1/(1+distancia_matriz)
    
    similaridade_index=pd.DataFrame(similaridade, index=dataframe.index, columns=dataframe.index)
    return similaridade_index

### Goleiros

In [34]:
# verificando os goleiros do Campeonato Brasileiro - Série A
jogadores_ok.loc[(jogadores_ok["posição"]=="Goalkeeper") & (jogadores_ok["id_liga"]==1396.0), ["id", "nome"]]

id                              nome
10    10012     Douglas Alan Schuck Friedrich
55    10372       Marcelo Lomba do Nascimento
108   10045                   Diego Cavalieri
109    2497   Roberto Júnior Fernández Torres
160    9902      Jaílson Marcelino dos Santos
162    2410         Weverton Pereira da Silva
220    9556                  Luan Polli Gomes
222    9557                           Mailson
265   41227             Muriel Gustavo Becker
267   10298                     Marcos Felipe
307    9944                  Tiago Luis Volpi
351   10151              Diego Alves Carreira
353  123759                        Hugo Souza
401   70366                              John
403    9979                        João Paulo
441   80184         Richard de Oliveira Costa
443    9903       Fernando Büttenbender Prass
480    9977         Vanderlei Farias da Silva
481   10476     Paulo Victor de Mileo Vidotti
523   10229                      Cássio Ramos
573   10547          Fernando Miguel Kaufmann
625   30763                           Jandrei
627   10112      Aderbar Melo dos Santos Neto
682    9942                              Jean
731    9336  Wilson Rodrigues de Moura Júnior
783   10335                             Tadeu
830   10408             Felipe Alves Raymundo
873    9449       Júlio César de Sousa Santos
874    9862                           Cleiton
912    9978      Everson Felipe Marques Pires
913   10081             Rafael Pires Monteiro

In [35]:
# definindo uma variável que exibe um Dataframe com as colunas em interesse
goleiros=jogadores_ok.loc[jogadores_ok["posição"]=="Goalkeeper", ["id", "idade", "jogos_escalados", "sofridos/min", "defesas/min", "passes/min"]]
goleiros.reset_index(drop=True, inplace=True)

In [36]:
# verificando o novo Dataframe
goleiros

id  idade  jogos_escalados  sofridos/min  defesas/min  passes/min
0     10012     33               24      0.017678     0.026756    0.172480
1     10372     35               28      0.009929     0.020254    0.255759
2     10045     39               20      0.020000     0.032778    0.245556
3      2497     33                8      0.009722     0.048611    0.211111
4      9902     40                4      0.013889     0.025000    0.233333
5      2410     34               23      0.009662     0.031401    0.295169
6      9556     28               22      0.013636     0.040404    0.253535
7      9557     25                7      0.015873     0.033333    0.117460
8     41227     34               19      0.011696     0.028070    0.183041
9     10298     25                9      0.014815     0.030864    0.238272
10     9944     31               29      0.010345     0.031801    0.339080
11    10151     36                9      0.014157     0.028314    0.261261
12   123759     22               13      0.016239     0.035897    0.297436
13    70366     25                6      0.012963     0.031481    0.268519
14     9979     26               18      0.014934     0.032258    0.301075
15    80184     30                7      0.009524     0.028571    0.236508
16     9903     43               22      0.016667     0.038889    0.224747
17     9977     37               21      0.009524     0.023810    0.188360
18    10476     34                7      0.007937     0.034921    0.144444
19    10229     34               24      0.013928     0.039461    0.299443
20    10547     36               26      0.014530     0.038889    0.219231
21    30763     28                4      0.008333     0.041667    0.344444
22    10112     31               24      0.009722     0.034722    0.365278
23     9942     26               27      0.011934     0.028395    0.341975
24     9336     37               29      0.014943     0.036782    0.287356
25    10335     29               26      0.017094     0.043590    0.318803
26    10408     33               28      0.010317     0.034524    0.395635
27     9449     37                6      0.016667     0.040741    0.283333
28     9862     24               22      0.012121     0.025253    0.244444
29     9978     31               17      0.015033     0.032026    0.356863
30    10081     32               10      0.009685     0.026634    0.259080
31     9662     36                6      0.013359     0.020992    0.248092
32    10334     28                5      0.022222     0.022222    0.135556
33     9410     25               19      0.012281     0.030994    0.197076
34     9373     35               30      0.007407     0.026296    0.279630
35    10186     33               33      0.005724     0.040067    0.337374
36    10080     41               32      0.009028     0.026042    0.223264
37     9512     24               26      0.010898     0.033130    0.295118
38    10548     38               12      0.006481     0.023148    0.279630
39    55105     25               16      0.011806     0.027083    0.338889
40     9731     29                8      0.010654     0.021309    0.324201
41    10110     22               18      0.011111     0.029630    0.243210
42   157459     29                6      0.016584     0.028192    0.225539
43     9692     23               18      0.016049     0.028395    0.233333
44     9694     26               13      0.011966     0.028205    0.409402
45     9335     30               31      0.009677     0.033333    0.287455
46     9627     37               11      0.022222     0.030303    0.244444
47    10228     25               16      0.010417     0.042361    0.182639
48     9377     25                4      0.022222     0.050000    0.227778
49    10509     32               24      0.014045     0.039794    0.228464
50    10511     38                9      0.014388     0.039568    0.250600
51     9763     29               10      0.013198     0.031472    0.278173
52     9227     28               1

In [37]:
# verificando valores que estão zerados / muito baixos
goleiros[["id", "sofridos/min"]].sort_values(by="sofridos/min")

id  sofridos/min
105   52604      0.000000
93    52694      0.000000
94    52978      0.000000
95    53134      0.000000
96   103883      0.000000
97    52487      0.000000
61     9229      0.000000
98    52484      0.000000
99    20820      0.000000
68    51822      0.000000
100   52415      0.000000
102   52773      0.000000
103   52951      0.000000
104  103777      0.000000
210   11984      0.000000
106  135332      0.000000
107   53017      0.000000
108   52527      0.000000
109   53094      0.000000
101   53019      0.000000
69    51652      0.000000
70    51733      0.000000
71    10297      0.000000
90    51622      0.000000
89    64500      0.000000
88    51533      0.000000
87     6291      0.000000
86    51325      0.000000
85    51794      0.000000
84    50077      0.000000
83     2967      0.000000
82    36494      0.000000
81    30493      0.000000
80    51475      0.000000
79    59542      0.000000
78    51385      0.000000
77    51844      0.000000
76     5536      0.000000
75    51762      0.000000
74    51353      0.000000
73    51592      0.000000
72   153381      0.000000
91    61895      0.000000
111   53133      0.000000
110   52952      0.000000
113   52733      0.000000
181   13631      0.000000
175   13234      0.000000
173   13130      0.000000
170   53093      0.000000
162   59466      0.000000
159   59649      0.000000
112  103789      0.000000
151   13352      0.000000
121   36095      0.000000
92    52654      0.000000
118   52451      0.000000
117   52914      0.000000
120   52774      0.000000
116   52913      0.000000
119  126664      0.000000
114   52562      0.000000
115   52810      0.000000
157   13664      0.000722
164   13457      0.000741
184   13760      0.000893
160   13758      0.001609
152    2481      0.001852
60    10442      0.001931
166   13495      0.002116
183   13726      0.002326
167   13568      0.002415
158   13665      0.002581
169   13598      0.003333
179   59574      0.003333
155   13163      0.003535
172   13129      0.004167
178   13235      0.004762
161   13456      0.004808
59    80173      0.004938
154   59882      0.004950
177   13277      0.005128
153   13496      0.005402
35    10186      0.005724
163   13381      0.005882
38    10548      0.006481
176   13316      0.006790
156   13762      0.007222
63    51681      0.007407
34     9373      0.007407
18    10476      0.007937
125   39868      0.008000
171   13696      0.008081
211   11346      0.008230
21    30763      0.008333
54   197017      0.008466
191   11438      0.008642
201   11657      0.008889
142   39614      0.008889
36    10080      0.009028
134   39777      0.009195
17     9977      0.009524
15    80184      0.009524
5      2410      0.009662
45     9335      0.009677
30    10081      0.009685
22    10112      0.009722
3      2497      0.009722
1     10372      0.009929
57     9303      0.010055
56    53932      0.010317
26    10408      0.010317
10     9944      0.010345
65     9193      0.010370
47    10228      0.010417
62     9478      0.010526
145   39745      0.010582
40     9731      0.010654
122   39744      0.010714
138   39499      0.010714
127   39893      0.010881
37     9512      0.010898
133   39778      0.010901
135   39440      0.011111
41    10110      0.011111
123    2416      0.011111
174   13698      0.011111
140   39407      0.011111
195   11818      0.011111
126   39474      0.011261
66     9589      0.011316
197   11819      0.011494
200   11755      0.011628
206   12063      0.011628
8     41227      0.011696
39    55105      0.011806
23     9942      0.011934
44     9694      0.011966
28     9862      0.012121
147   39473      0.012121
190    6699      0.012222
33     9410      0.012281
64     9194      0.012500
198    6423      0.012500
193   11379      0.012821
139   39406      0.012963
207   11557      0.012963
13    70366      0.012963
205   11525      0.013187
51     9763      0.013198
194   11440      0.013248
146   39681      0.013333
186   11467      0.013333
208  

In [38]:
# verificando dados de exemplo de goleiros que possui valor zerado / muito baixo
jogadores_ok.loc[(jogadores_ok["id"]==9229) | (jogadores_ok["id"]==51592) | (jogadores_ok["id"]==13664), :]

id                             nome     posição  idade  \
1823   9229  Matheus Vinicius Matos Nogueira  Goalkeeper     23   
2144  51592     Kevin Emiliano Dawson Blanco  Goalkeeper     28   
4077  13664                         G. Ortiz  Goalkeeper     31   

     data_nascimento nacionalidade         time  id_liga                liga  \
1823      20/09/1997        Brazil       Cuiaba   1397.0  Brasileiro Série B   
2144      08/02/1992       Uruguay      Penarol   1375.0        Liga Uruguai   
4077      25/03/1989      Paraguay  Once Caldas   1326.0       Liga Colômbia   

      total_chutes  chutes_no_gol  gols  gols_sofridos  assistências  defesas  \
1823             0              0     0              0             0        1   
2144             0              0     0              0             0        0   
4077             0              0     0              1             0        8   

      total_passes  passes_chaves  %passes_certos  desarmes  dribles_tentados  \
1823             6              0               1         0                 0   
2144             0              0               0         0                 0   
4077            47              0              63         0                 0   

      dribles_realizados  faltas_sofridas  faltas_cometidas  total_jogos  \
1823                   0                0                 0            5   
2144                   0                0                 0           20   
4077                   0                0                 0           16   

      minutos_jogados  jogos_escalados  chutes/min  chutes_no_gol/min  \
1823              378                4         0.0                0.0   
2144             1800               20         0.0                0.0   
4077             1385               16         0.0                0.0   

      gols/min  sofridos/min  assistências/min  defesas/min  passes/min  \
1823       0.0      0.000000               0.0     0.002646    0.015873   
2144       0.0      0.000000               0.0     0.000000    0.000000   
4077       0.0      0.000722               0.0     0.005776    0.033935   

      passes_chaves/min  desarmes/min  dribles_tentados/min  \
1823                0.0           0.0                   0.0   
2144                0.0           0.0                   0.0   
4077                0.0           0.0                   0.0   

      dribles_realizados/min  faltas_sofridas/min  faltas_cometidas/min  
1823                     0.0                  0.0                   0.0  
2144                     0.0                  0.0                   0.0  
4077                     0.0                  0.0                   0.0

> Goleiros que possuem valores zerados apresentam dados incorretos. Verifica-se no caso acima que o goleiro Kevin Emiliano jogou 1800 minutos, porém não foi registrado nenhum gol sofrido nem defesa, por exemplo.

> Já o goleiro Matheus Vinicius não sofreu nenhum gol pelo fato de ter jogado apenas 378 minutos, por isso apresenta valor zerado em defesas por minuto.

> Por fim, o goleiro G. Ortiz possui um valor de defesas por minuto muito alto, o que significa que o número de gols sofridos registrado está errado.

> **Apesar da tabela possuir valores errados, todos eles serão considerados com o objetivo de fazer a comparação, até porque não dá para identificar todos os erros. A ideia principal é criar uma base para que seja possível fazer o aplicativo, e futuramente buscar uma base de dados mais confiável para ser aplicada.**

In [39]:
# verificando o DataFrame
goleiros

id  idade  jogos_escalados  sofridos/min  defesas/min  passes/min
0     10012     33               24      0.017678     0.026756    0.172480
1     10372     35               28      0.009929     0.020254    0.255759
2     10045     39               20      0.020000     0.032778    0.245556
3      2497     33                8      0.009722     0.048611    0.211111
4      9902     40                4      0.013889     0.025000    0.233333
5      2410     34               23      0.009662     0.031401    0.295169
6      9556     28               22      0.013636     0.040404    0.253535
7      9557     25                7      0.015873     0.033333    0.117460
8     41227     34               19      0.011696     0.028070    0.183041
9     10298     25                9      0.014815     0.030864    0.238272
10     9944     31               29      0.010345     0.031801    0.339080
11    10151     36                9      0.014157     0.028314    0.261261
12   123759     22               13      0.016239     0.035897    0.297436
13    70366     25                6      0.012963     0.031481    0.268519
14     9979     26               18      0.014934     0.032258    0.301075
15    80184     30                7      0.009524     0.028571    0.236508
16     9903     43               22      0.016667     0.038889    0.224747
17     9977     37               21      0.009524     0.023810    0.188360
18    10476     34                7      0.007937     0.034921    0.144444
19    10229     34               24      0.013928     0.039461    0.299443
20    10547     36               26      0.014530     0.038889    0.219231
21    30763     28                4      0.008333     0.041667    0.344444
22    10112     31               24      0.009722     0.034722    0.365278
23     9942     26               27      0.011934     0.028395    0.341975
24     9336     37               29      0.014943     0.036782    0.287356
25    10335     29               26      0.017094     0.043590    0.318803
26    10408     33               28      0.010317     0.034524    0.395635
27     9449     37                6      0.016667     0.040741    0.283333
28     9862     24               22      0.012121     0.025253    0.244444
29     9978     31               17      0.015033     0.032026    0.356863
30    10081     32               10      0.009685     0.026634    0.259080
31     9662     36                6      0.013359     0.020992    0.248092
32    10334     28                5      0.022222     0.022222    0.135556
33     9410     25               19      0.012281     0.030994    0.197076
34     9373     35               30      0.007407     0.026296    0.279630
35    10186     33               33      0.005724     0.040067    0.337374
36    10080     41               32      0.009028     0.026042    0.223264
37     9512     24               26      0.010898     0.033130    0.295118
38    10548     38               12      0.006481     0.023148    0.279630
39    55105     25               16      0.011806     0.027083    0.338889
40     9731     29                8      0.010654     0.021309    0.324201
41    10110     22               18      0.011111     0.029630    0.243210
42   157459     29                6      0.016584     0.028192    0.225539
43     9692     23               18      0.016049     0.028395    0.233333
44     9694     26               13      0.011966     0.028205    0.409402
45     9335     30               31      0.009677     0.033333    0.287455
46     9627     37               11      0.022222     0.030303    0.244444
47    10228     25               16      0.010417     0.042361    0.182639
48     9377     25                4      0.022222     0.050000    0.227778
49    10509     32               24      0.014045     0.039794    0.228464
50    10511     38                9      0.014388     0.039568    0.250600
51     9763     29               10      0.013198     0.031472    0.278173
52     9227     28               1

#### Similaridade dos goleiros

In [40]:
# importando biblioteca de normalização
from sklearn.preprocessing import StandardScaler
scaler=StandardScaler()

In [41]:
# checando as colunas do Dataframe criado
goleiros.columns

Index(['id', 'idade', 'jogos_escalados', 'sofridos/min', 'defesas/min',
       'passes/min'],
      dtype='object')

In [42]:
# criando uma variável que exibe a normalização de todos os dados menos o id
scaled_goleiros=scaler.fit_transform(goleiros.drop(columns="id"))

In [43]:
# renomeando a variável de goleiros para seus respectivos id e os valores normalizados
goleiros=goleiros[["id"]].merge(pd.DataFrame(scaled_goleiros, columns=goleiros[["idade", "jogos_escalados", "sofridos/min", "defesas/min", "passes/min"]].columns), on=goleiros.index).drop("key_0", axis=1)                             

In [44]:
# verificando o Dataframe normalizado
goleiros

id     idade  jogos_escalados  sofridos/min  defesas/min  passes/min
0     10012  0.559365         1.128306      1.394275     0.497181    0.185252
1     10372  0.932275         1.636973      0.248018     0.063661    0.859988
2     10045  1.678096         0.619638      1.737737     0.898712    0.777322
3      2497  0.559365        -0.906364      0.217504     1.954451    0.498249
4      9902  1.864551        -1.415031      0.833814     0.380104    0.678296
5      2410  0.745820         1.001139      0.208572     0.806909    1.179298
6      9556 -0.372910         0.873972      0.796462     1.407218    0.841976
7      9557 -0.932275        -1.033530      1.127296     0.935756   -0.260523
8     41227  0.745820         0.492472      0.509440     0.584818    0.270820
9     10298 -0.932275        -0.779197      0.970772     0.771118    0.718307
10     9944  0.186455         1.764140      0.309596     0.833567    1.535074
11    10151  1.118730        -0.779197      0.873474     0.601078    0.904572
12   123759 -1.491641        -0.270529      1.181477     1.106726    1.197664
13    70366 -0.932275        -1.160697      0.696857     0.812278    0.963372
14     9979 -0.745820         0.365305      0.988444     0.864059    1.227151
15    80184  0.000000        -1.033530      0.188156     0.618241    0.704017
16     9903  2.423916         0.873972      1.244688     1.306190    0.608732
17     9977  1.305186         0.746805      0.188156     0.300725    0.313914
18    10476  0.745820        -1.033530     -0.046629     1.041594   -0.041894
19    10229  0.745820         1.128306      0.839537     1.344369    1.213925
20    10547  1.118730         1.382639      0.928631     1.306190    0.564035
21    30763 -0.372910        -1.415031      0.012067     1.491408    1.578534
22    10112  0.186455         1.128306      0.217504     1.028365    1.747329
23     9942 -0.745820         1.509806      0.544681     0.606481    1.558529
24     9336  1.305186         1.764140      0.989663     1.165681    1.115998
25    10335 -0.186455         1.382639      1.307899     1.619635    1.370787
26    10408  0.559365         1.636973      0.305548     1.015135    1.993287
27     9449  1.305186        -1.160697      1.244688     1.429669    1.083403
28     9862 -1.118730         0.873972      0.572349     0.396942    0.768320
29     9978  0.186455         0.238138      1.002998     0.848595    1.679149
30    10081  0.372910        -0.652030      0.212032     0.489082    0.886899
31     9662  1.118730        -1.160697      0.755403     0.112882    0.797870
32    10334 -0.372910        -1.287864      2.066436     0.194887   -0.113913
33     9410 -0.932275         0.492472      0.595940     0.779784    0.384534
34     9373  0.932275         1.891307     -0.124891     0.466539    1.053395
35    10186  0.559365         2.272807     -0.373905     1.384767    1.521246
36    10080  2.051006         2.145640      0.114785     0.449561    0.596712
37     9512 -1.118730         1.382639      0.391418     0.922192    1.178882
38    10548  1.491641        -0.397696     -0.261849     0.256626    1.053395
39    55105 -0.932275         0.110971      0.525659     0.519017    1.533522
40     9731 -0.186455        -0.906364      0.355400     0.133994    1.414518
41    10110 -1.491641         0.365305      0.422941     0.688800    0.758317
42   157459 -0.186455        -1.160697      1.232423     0.592966    0.615145
43     9692 -1.305186         0.365305      1.153383     0.606481    0.678296
44     9694 -0.745820        -0.270529      0.549363     0.593816    2.104827
45     9335  0.000000         2.018474      0.210877     0.935756    1.116799
46     9627  1.305186        -0.524863      2.066436     0.733701    0.768320
47    10228 -0.932275         0.110971      0.320222     1.537712    0.267563
48     9377 -0.932275        -1.415031      2.066436     2.047059    0.633284
49    10509  0.372910         1.128306      0.856897     1.366542    0.638848
50    10511  1.491641        -0

In [45]:
# transformando o id em index
goleiros.index=goleiros.id
goleiros.drop(columns="id", inplace=True)

In [46]:
# verificando o Dataframe
goleiros

idade  jogos_escalados  sofridos/min  defesas/min  passes/min
id                                                                      
10012   0.559365         1.128306      1.394275     0.497181    0.185252
10372   0.932275         1.636973      0.248018     0.063661    0.859988
10045   1.678096         0.619638      1.737737     0.898712    0.777322
2497    0.559365        -0.906364      0.217504     1.954451    0.498249
9902    1.864551        -1.415031      0.833814     0.380104    0.678296
2410    0.745820         1.001139      0.208572     0.806909    1.179298
9556   -0.372910         0.873972      0.796462     1.407218    0.841976
9557   -0.932275        -1.033530      1.127296     0.935756   -0.260523
41227   0.745820         0.492472      0.509440     0.584818    0.270820
10298  -0.932275        -0.779197      0.970772     0.771118    0.718307
9944    0.186455         1.764140      0.309596     0.833567    1.535074
10151   1.118730        -0.779197      0.873474     0.601078    0.904572
123759 -1.491641        -0.270529      1.181477     1.106726    1.197664
70366  -0.932275        -1.160697      0.696857     0.812278    0.963372
9979   -0.745820         0.365305      0.988444     0.864059    1.227151
80184   0.000000        -1.033530      0.188156     0.618241    0.704017
9903    2.423916         0.873972      1.244688     1.306190    0.608732
9977    1.305186         0.746805      0.188156     0.300725    0.313914
10476   0.745820        -1.033530     -0.046629     1.041594   -0.041894
10229   0.745820         1.128306      0.839537     1.344369    1.213925
10547   1.118730         1.382639      0.928631     1.306190    0.564035
30763  -0.372910        -1.415031      0.012067     1.491408    1.578534
10112   0.186455         1.128306      0.217504     1.028365    1.747329
9942   -0.745820         1.509806      0.544681     0.606481    1.558529
9336    1.305186         1.764140      0.989663     1.165681    1.115998
10335  -0.186455         1.382639      1.307899     1.619635    1.370787
10408   0.559365         1.636973      0.305548     1.015135    1.993287
9449    1.305186        -1.160697      1.244688     1.429669    1.083403
9862   -1.118730         0.873972      0.572349     0.396942    0.768320
9978    0.186455         0.238138      1.002998     0.848595    1.679149
10081   0.372910        -0.652030      0.212032     0.489082    0.886899
9662    1.118730        -1.160697      0.755403     0.112882    0.797870
10334  -0.372910        -1.287864      2.066436     0.194887   -0.113913
9410   -0.932275         0.492472      0.595940     0.779784    0.384534
9373    0.932275         1.891307     -0.124891     0.466539    1.053395
10186   0.559365         2.272807     -0.373905     1.384767    1.521246
10080   2.051006         2.145640      0.114785     0.449561    0.596712
9512   -1.118730         1.382639      0.391418     0.922192    1.178882
10548   1.491641        -0.397696     -0.261849     0.256626    1.053395
55105  -0.932275         0.110971      0.525659     0.519017    1.533522
9731   -0.186455        -0.906364      0.355400     0.133994    1.414518
10110  -1.491641         0.365305      0.422941     0.688800    0.758317
157459 -0.186455        -1.160697      1.232423     0.592966    0.615145
9692   -1.305186         0.365305      1.153383     0.606481    0.678296
9694   -0.745820        -0.270529      0.549363     0.593816    2.104827
9335    0.000000         2.018474      0.210877     0.935756    1.116799
9627    1.305186        -0.524863      2.066436     0.733701    0.768320
10228  -0.932275         0.110971      0.320222     1.537712    0.267563
9377   -0.932275        -1.415031      2.066436     2.047059    0.633284
10509   0.372910         1.128306      0.856897     1.366542    0.638848
10511   1.491641        -0.779197      0.907713     1.351495    0.818189
9763   -0.186455        -0.652030      0.731617     0.811651    1.041590
9227   -0.372910         0.365305      0.806995     0.802283    1.225

In [47]:
# criando uma variável que exibe um Dataframe que mostra os id similares de cada id
similaridade_goleiros=criar_similaridade(goleiros)

In [48]:
# verificando o novo Dataframe
similaridade_goleiros

id        10012     10372     10045     2497      9902      2410      9556    \
id                                                                             
10012   1.000000  0.394543  0.406027  0.264326  0.252726  0.385503  0.385048   
10372   0.394543  1.000000  0.320024  0.237438  0.234467  0.488660  0.323029   
10045   0.406027  0.320024  1.000000  0.273109  0.303421  0.347587  0.300471   
2497    0.264326  0.237438  0.273109  1.000000  0.312210  0.299814  0.313582   
9902    0.252726  0.234467  0.303421  0.312210  1.000000  0.262254  0.229051   
2410    0.385503  0.488660  0.347587  0.299814  0.262254  1.000000  0.409028   
9556    0.385048  0.323029  0.300471  0.313582  0.229051  0.409028  1.000000   
9557    0.269317  0.214581  0.231678  0.316068  0.247500  0.241267  0.298960   
41227   0.473440  0.412306  0.376217  0.333213  0.304778  0.474764  0.388282   
10298   0.283267  0.237063  0.246155  0.326444  0.256526  0.277476  0.348456   
9944    0.343428  0.439630  0.287315  0.243897  0.210084  0.495996  0.405368   
10151   0.314499  0.280883  0.361437  0.375920  0.492465  0.337055  0.296722   
123759  0.266117  0.227066  0.228739  0.277956  0.213986  0.265414  0.368909   
70366   0.253675  0.223460  0.230785  0.332219  0.259136  0.263971  0.312692   
9979    0.342952  0.294175  0.278597  0.296396  0.235460  0.357035  0.516252   
80184   0.278300  0.257251  0.260792  0.405036  0.330694  0.309918  0.313635   
9903    0.322892  0.300715  0.492895  0.259608  0.280436  0.320849  0.260153   
9977    0.402064  0.468267  0.361163  0.288894  0.298131  0.459547  0.315538   
10476   0.272430  0.249762  0.267772  0.472559  0.361172  0.294273  0.281483   
10229   0.407254  0.391840  0.396130  0.299971  0.250661  0.543698  0.452803   
10547   0.458718  0.403020  0.429716  0.281999  0.247311  0.455330  0.383299   
30763   0.222733  0.213142  0.221747  0.385009  0.264569  0.264204  0.283230   
10112   0.326733  0.385827  0.294252  0.278721  0.229846  0.544318  0.435808   
9942    0.321138  0.341966  0.252254  0.235063  0.198624  0.374877  0.429707   
9336    0.390438  0.415111  0.402999  0.247060  0.229645  0.438635  0.339673   
10335   0.355255  0.306458  0.306586  0.262934  0.209287  0.369151  0.515997   
10408   0.309416  0.395710  0.292869  0.244058  0.214378  0.482061  0.359598   
9449    0.267552  0.232679  0.335373  0.397686  0.426232  0.282401  0.271283   
9862    0.335973  0.308730  0.244996  0.254173  0.209624  0.334394  0.439036   
9978    0.350304  0.323574  0.341258  0.316044  0.277204  0.428573  0.428918   
10081   0.307023  0.294562  0.292897  0.392389  0.356819  0.363786  0.331368   
9662    0.281985  0.259705  0.307631  0.329307  0.542221  0.294501  0.260677   
10334   0.269601  0.207818  0.247512  0.262370  0.271447  0.224084  0.253520   
9410    0.352093  0.296190  0.257641  0.294883  0.225745  0.337385  0.487911   
9373    0.339538  0.611551  0.291530  0.234930  0.217996  0.491536  0.319968   
10186   0.274164  0.361757  0.247609  0.225267  0.188712  0.390213  0.316643   
10080   0.307564  0.430761  0.302115  0.211947  0.215568  0.348679  0.250686   
9512    0.307864  0.306270  0.236876  0.243858  0.193074  0.343004  0.464618   
10548   0.277895  0.313676  0.297864  0.319212  0.385990  0.363998  0.266411   
55105   0.292816  0.281058  0.247360  0.281929  0.231690  0.335659  0.400774   
9731    0.268552  0.260674  0.253117  0.315001  0.302894  0.308222  0.301884   
10110   0.288243  0.262123  0.224847  0.267017  0.206842  0.296167  0.404189   
157459  0.289625  0.236799  0.273990  0.347740  0.320875  0.274934  0.307342   
9692    0.323532  0.264019  0.245932  0.263418  0.214752  0.293494  0.418969   
9694    0.260385  0.256796  0.240847  0.279421  0.237693  0.312051  0.340052   
9335    0.346219  0.424194  0.271214  0.237672  0.199232  0.440618  0.408903   
9627    0.329604  0.254545  0.442735  0.295507  0.375847  0.285429  0.276532   
10228   0.298877  0.257274  0.242568  0.348416  0.224566  0.309312  0.452088   


In [49]:
# exemplo: goleiros similares ao Diego Alves do Flamengo
similaridade_goleiros[10151].sort_values(ascending=False)

id
10151     1.000000
11557     0.669016
9662      0.609871
10511     0.542563
11593     0.529150
39474     0.519402
11467     0.509118
10081     0.497184
9449      0.495501
9902      0.492465
11754     0.482472
12063     0.476345
11755     0.470507
9626      0.468879
9194      0.460804
9375      0.450530
9627      0.444740
10548     0.433025
13279     0.430747
51681     0.430030
9763      0.426853
80184     0.425269
13279     0.420147
9478      0.411367
157459    0.410530
10476     0.406019
5871      0.405646
11594     0.401741
41227     0.397854
39809     0.396057
6699      0.390734
9731      0.389038
67911     0.387664
39406     0.386647
39891     0.384964
9978      0.383598
39373     0.378855
2497      0.375920
13698     0.375699
11497     0.364844
10045     0.361437
11525     0.360237
9977      0.356009
2416      0.346485
11655     0.345363
11438     0.345070
2415      0.343728
39614     0.343044
9227      0.342570
2410      0.337055
39745     0.332759
39587     0.332100
196670   

In [50]:
# observando goleiros similares a Diego Alves
jogadores_ok.loc[(jogadores_ok["id"]==10151) | (jogadores_ok["id"]==39474) | (jogadores_ok["id"]==9627) | (jogadores_ok["id"]==10511) | (jogadores_ok["id"]==11557), :]

id                         nome     posição  idade data_nascimento  \
351   10151         Diego Alves Carreira  Goalkeeper     36      24/06/1985   
1421   9627             Luiz Silva Filho  Goalkeeper     37      07/02/1983   
1493  10511   Gledson Ribeiro dos Santos  Goalkeeper     38      06/02/1983   
3398  39474  Exar Javier Rosales Sánchez  Goalkeeper     36      20/05/1984   
5199  11557         Jorge Luis Deschamps  Goalkeeper     37      13/05/1984   

     nacionalidade         time  id_liga                liga  total_chutes  \
351         Brazil     Flamengo   1396.0  Brasileiro Série A             0   
1421        Brazil        Oeste   1397.0  Brasileiro Série B             0   
1493        Brazil         Avai   1397.0  Brasileiro Série B             0   
3398          Peru  Ayacucho FC   1341.0           Liga Peru             0   
5199         Chile     Cobresal   1342.0          Liga Chile             0   

      chutes_no_gol  gols  gols_sofridos  assistências  defesas  total_passes  \
351               0     0             11             0       22           203   
1421              0     0             22             0       30           242   
1493              0     0             12             0       33           209   
3398              0     0             10             1       21           150   
5199              0     0              7             0       15           153   

      passes_chaves  %passes_certos  desarmes  dribles_tentados  \
351               0              56         0                 0   
1421              0              43         0                 0   
1493              0              12         0                 1   
3398              1              57         0                 1   
5199              0              18         0                 1   

      dribles_realizados  faltas_sofridas  faltas_cometidas  total_jogos  \
351                    0                3                 0            9   
1421                   0                1                 0           11   
1493                   1                5                 0           10   
3398                   1                3                 2           10   
5199                   1                1                 0            6   

      minutos_jogados  jogos_escalados  chutes/min  chutes_no_gol/min  \
351               777                9         0.0                0.0   
1421              990               11         0.0                0.0   
1493              834                9         0.0                0.0   
3398              888               10         0.0                0.0   
5199              540                6         0.0                0.0   

      gols/min  sofridos/min  assistências/min  defesas/min  passes/min  \
351        0.0      0.014157          0.000000     0.028314    0.261261   
1421       0.0      0.022222          0.000000     0.030303    0.244444   
1493       0.0      0.014388          0.000000     0.039568    0.250600   
3398       0.0      0.011261          0.001126     0.023649    0.168919   
5199       0.0      0.012963          0.000000     0.027778    0.283333   

      passes_chaves/min  desarmes/min  dribles_tentados/min  \
351            0.000000           0.0              0.000000   
1421           0.000000           0.0              0.000000   
1493           0.000000           0.0              0.001199   
3398           0.001126           0.0              0.001126   
5199           0.000000           0.0              0.001852   

      dribles_realizados/min  faltas_sofridas/min  faltas_cometidas/min  
351                 0.000000             0.003861              0.000000  
1421                0.000000             0.001010              0.000000  
1493                0.001199             0.005995              0.000000  
3398                0.001126             0.003378              0.002252  
5199                0.001852             0.001852              0.000000

### Defensores

In [51]:
# definindo uma variável que exibe um Dataframe com as colunas em interesse
defesa=jogadores_ok.loc[jogadores_ok["posição"]=="Defender", ["id", "idade", "jogos_escalados", "passes/min", "desarmes/min", "faltas_sofridas/min", "faltas_cometidas/min"]]
defesa.reset_index(drop=True, inplace=True)

In [52]:
# verificando o novo Dataframe
defesa

id  idade  jogos_escalados  passes/min  desarmes/min  \
0    197051     22                6    0.388769      0.021598   
1     10015     35               22    0.374187      0.022011   
2     10376     27                4    0.538095      0.002381   
3     10481     23               19    0.519556      0.026270   
4      9912     26               22    0.411561      0.011364   
..      ...    ...              ...         ...           ...   
925   11481     26                5    0.744035      0.017354   
926   11356     24               17    0.423253      0.014370   
927   11349     31                6    0.455645      0.012097   
928    6504     32               14    0.370796      0.013126   
929   11365     25                7    0.363492      0.004762   

     faltas_sofridas/min  faltas_cometidas/min  
0               0.012959              0.021598  
1               0.012506              0.017009  
2               0.002381              0.007143  
3               0.016929              0.012843  
4               0.004447              0.002964  
..                   ...                   ...  
925             0.010846              0.032538  
926             0.006532              0.011757  
927             0.012097              0.010081  
928             0.004102              0.009844  
929             0.004762              0.012698  

[930 rows x 7 columns]

#### Similaridade dos defensores

In [53]:
# criando uma variável que exibe a normalização de todos os dados menos o id
scaled_defesa=scaler.fit_transform(defesa.drop(columns="id"))

In [54]:
# renomeando a variável de defesa para seus respectivos id e os valores normalizados
defesa=defesa[["id"]].merge(pd.DataFrame(scaled_defesa, columns=defesa[["idade", "jogos_escalados", "passes/min", "desarmes/min", "faltas_sofridas/min", "faltas_cometidas/min"]].columns), on=defesa.index).drop("key_0", axis=1)

In [55]:
# transformando o id em index
defesa.index=defesa.id
defesa.drop(columns="id", inplace=True)

In [56]:
# criando uma variável que exibe um Dataframe que mostra os id similares de cada id
similaridade_defesa=criar_similaridade(defesa)

In [57]:
# verificando o novo Dataframe
similaridade_defesa

id        197051    10015     10376     10481     9912      10022     10024   \
id                                                                             
197051  1.000000  0.210988  0.231014  0.285185  0.206244  0.191004  0.204534   
10015   0.210988  1.000000  0.185954  0.264241  0.249680  0.292646  0.277858   
10376   0.231014  0.185954  1.000000  0.192731  0.243669  0.247375  0.287268   
10481   0.285185  0.264241  0.192731  1.000000  0.260267  0.214904  0.202514   
9912    0.206244  0.249680  0.243669  0.260267  1.000000  0.370899  0.270803   
...          ...       ...       ...       ...       ...       ...       ...   
11481   0.302848  0.194081  0.209466  0.213534  0.171397  0.177407  0.189723   
11356   0.288575  0.264757  0.273077  0.326657  0.407805  0.309799  0.268468   
11349   0.281805  0.250605  0.329343  0.241838  0.244832  0.260483  0.354128   
6504    0.236532  0.306351  0.298137  0.238620  0.336104  0.427240  0.494125   
11365   0.288336  0.206879  0.437934  0.220701  0.267161  0.259880  0.283594   

id        9953      10013     10377     10378     10157     10382     10384   \
id                                                                             
197051  0.226704  0.332627  0.192618  0.231512  0.366129  0.250144  0.418378   
10015   0.220551  0.251266  0.387630  0.264171  0.306846  0.301688  0.211509   
10376   0.335595  0.203904  0.194467  0.192302  0.212207  0.289341  0.197397   
10481   0.186606  0.276348  0.255966  0.242063  0.320698  0.226987  0.360407   
9912    0.211251  0.218599  0.290393  0.189684  0.218500  0.260827  0.205674   
...          ...       ...       ...       ...       ...       ...       ...   
11481   0.244883  0.293083  0.188090  0.210890  0.280308  0.249699  0.237384   
11356   0.241876  0.300387  0.280409  0.214068  0.286804  0.310742  0.270014   
11349   0.354612  0.234221  0.234081  0.306195  0.317258  0.358336  0.249335   
6504    0.321758  0.253720  0.292335  0.230803  0.277432  0.459376  0.215892   
11365   0.318889  0.244228  0.209757  0.206235  0.249888  0.319548  0.229983   

id        2470      51767     9907      10052     146378    10338     1058    \
id                                                                             
197051  0.357616  0.331862  0.251179  0.223565  0.224112  0.272714  0.253361   
10015   0.260351  0.205764  0.372303  0.272098  0.156947  0.290800  0.372747   
10376   0.208948  0.299086  0.207664  0.225808  0.162246  0.209532  0.255114   
10481   0.315441  0.250071  0.432611  0.299731  0.174137  0.517904  0.296422   
9912    0.209404  0.225019  0.325191  0.429271  0.144443  0.317115  0.317474   
...          ...       ...       ...       ...       ...       ...       ...   
11481   0.240918  0.231984  0.203496  0.195914  0.209130  0.218593  0.226214   
11356   0.263561  0.281333  0.369946  0.441519  0.172457  0.414794  0.368373   
11349   0.321290  0.407041  0.272739  0.237729  0.185872  0.251021  0.359275   
6504    0.254195  0.262292  0.309269  0.300851  0.157561  0.274261  0.409548   
11365   0.239502  0.348294  0.236171  0.259032  0.188440  0.242786  0.289276   

id        9418      10059     9913      2502      9916      9917      10385   \
id                                                                             
197051  0.294375  0.206232  0.251876  0.256234  0.245141  0.338756  0.375398   
10015   0.289527  0.256329  0.282974  0.286034  0.308727  0.295203  0.230536   
10376   0.227418  0.200033  0.304706  0.272190  0.184743  0.267094  0.353161   
10481   0.232271  0.307299  0.286557  0.288611  0.270436  0.319869  0.269262   
9912    0.203878  0.381540  0.415705  0.314844  0.201951  0.261557  0.255351   
...          ...       ...       ...       ...       ...       ...       ...   
11481   0.270102  0.179145  0.217357  0.239892  0.229430  0.265527  0.271910   
11356   0.252241  0.356433  0.481767  0.387075  0.235719  0.345815  0.350901   
11349   0.352454  0.215550  0.331926  0.323499  0.252961  0.420778  0.452478 

### Meias

In [58]:
# definindo uma variável que exibe um Dataframe com as colunas em interesse
meias=jogadores_ok.loc[jogadores_ok["posição"]=="Midfielder", ["id", "idade", "%passes_certos", "chutes/min", "gols/min", "assistências/min", "passes/min", "passes_chaves/min", "desarmes/min", "dribles_tentados/min", "dribles_realizados/min"]]
meias.reset_index(drop=True, inplace=True)

In [59]:
# verificando o novo Dataframe
meias

id  idade  %passes_certos  chutes/min  gols/min  assistências/min  \
0     10094     33              27    0.020140  0.003503          0.000876   
1     64032     23              15    0.004073  0.000000          0.000000   
2     10167     25              46    0.006436  0.000000          0.000805   
3     10031     27              55    0.007081  0.000000          0.001517   
4      9888     36              29    0.008838  0.000000          0.000000   
..      ...    ...             ...         ...       ...               ...   
959   11359     34              49    0.008399  0.004941          0.001976   
960   11360     33              33    0.013538  0.000000          0.006318   
961   11364     36              45    0.010699  0.002517          0.001888   
962  195121     21              28    0.021322  0.002132          0.004264   
963   11368     30              41    0.010714  0.004762          0.001190   

     passes/min  passes_chaves/min  desarmes/min  dribles_tentados/min  \
0      0.289842           0.009632      0.007881              0.027145   
1      0.431772           0.004073      0.024440              0.010183   
2      0.471440           0.004023      0.034594              0.008850   
3      0.637835           0.009611      0.026302              0.017704   
4      0.488636           0.006313      0.020202              0.010101   
..          ...                ...           ...                   ...   
959    0.471838           0.015316      0.016304              0.013834   
960    0.422383           0.027978      0.010830              0.018051   
961    0.417873           0.018880      0.007552              0.014475   
962    0.339019           0.012793      0.029851              0.008529   
963    0.284524           0.015476      0.008333              0.017857   

     dribles_realizados/min  
0                  0.016637  
1                  0.006110  
2                  0.005632  
3                  0.014669  
4                  0.005051  
..                      ...  
959                0.008893  
960                0.010830  
961                0.008181  
962                0.006397  
963                0.015476  

[964 rows x 11 columns]

#### Similaridade dos meias

In [60]:
# criando uma variável que exibe a normalização de todos os dados menos o id
scaled_meias=scaler.fit_transform(meias.drop(columns="id"))

In [61]:
# renomeando a variável de meias para seus respectivos id e os valores normalizados
meias=meias[["id"]].merge(pd.DataFrame(scaled_meias, columns=meias[["idade", "%passes_certos", "chutes/min", "gols/min", "assistências/min", "passes/min", "passes_chaves/min", "desarmes/min", "dribles_tentados/min", "dribles_realizados/min"]].columns), on=meias.index).drop("key_0", axis=1)

In [62]:
# transformando o id em index
meias.index=meias.id
meias.drop(columns="id", inplace=True)

In [63]:
# criando uma variável que exibe um Dataframe que mostra os id similares de cada id
similaridade_meias=criar_similaridade(meias)

In [64]:
# verificando o novo Dataframe
similaridade_meias

id        10094     64032     10167     10031     9888      80537     10035   \
id                                                                             
10094   1.000000  0.174938  0.170168  0.192036  0.203102  0.251301  0.141037   
64032   0.174938  1.000000  0.361129  0.258330  0.259443  0.236575  0.116655   
10167   0.170168  0.361129  1.000000  0.322132  0.259234  0.200224  0.120015   
10031   0.192036  0.258330  0.322132  1.000000  0.244544  0.206600  0.139663   
9888    0.203102  0.259443  0.259234  0.244544  1.000000  0.196757  0.114103   
...          ...       ...       ...       ...       ...       ...       ...   
11359   0.247650  0.169225  0.179998  0.201392  0.196337  0.184046  0.148014   
11360   0.135859  0.119279  0.127335  0.148442  0.127364  0.126513  0.209148   
11364   0.264263  0.178021  0.184280  0.216877  0.233759  0.199322  0.153487   
195121  0.163255  0.163673  0.183221  0.186309  0.147968  0.173176  0.173863   
11368   0.330544  0.176670  0.173614  0.194693  0.187379  0.236488  0.151109   

id        10320     10391     47577     10387     10388     21093     197533  \
id                                                                             
10094   0.190675  0.175286  0.183053  0.244707  0.196751  0.261139  0.176089   
64032   0.170554  0.165087  0.244021  0.206446  0.149631  0.261460  0.283935   
10167   0.169468  0.164098  0.261417  0.229976  0.152570  0.271755  0.262053   
10031   0.208515  0.193677  0.247999  0.288305  0.182380  0.301539  0.268523   
9888    0.169394  0.225458  0.334007  0.246618  0.147799  0.239547  0.192119   
...          ...       ...       ...       ...       ...       ...       ...   
11359   0.216380  0.178647  0.217845  0.291461  0.155628  0.216414  0.157459   
11360   0.158994  0.134542  0.121354  0.188212  0.118810  0.145756  0.119291   
11364   0.240466  0.224929  0.224333  0.378562  0.153445  0.237427  0.162449   
195121  0.172562  0.128708  0.147184  0.218800  0.136183  0.219836  0.163803   
11368   0.226099  0.172163  0.191767  0.253193  0.175464  0.239856  0.172725   

id        133185    125677    10393     80166     10395     9882      10064   \
id                                                                             
10094   0.170137  0.160304  0.185910  0.174878  0.177565  0.200843  0.207248   
64032   0.409053  0.381031  0.368706  0.110551  0.254413  0.333261  0.170462   
10167   0.327915  0.336194  0.326058  0.112344  0.278158  0.355366  0.156751   
10031   0.276608  0.268475  0.232217  0.115479  0.283436  0.428341  0.180898   
9888    0.224901  0.234825  0.287204  0.116990  0.347208  0.264180  0.153042   
...          ...       ...       ...       ...       ...       ...       ...   
11359   0.162567  0.165812  0.198902  0.178947  0.203534  0.201916  0.145964   
11360   0.116718  0.115362  0.117356  0.114879  0.136648  0.150544  0.131250   
11364   0.168038  0.171444  0.195259  0.153029  0.238420  0.222546  0.163239   
195121  0.162462  0.151735  0.164433  0.144305  0.158378  0.199826  0.153733   
11368   0.169424  0.167760  0.193932  0.180614  0.182934  0.203541  0.179310   

id        6857      25352     13369     180236    12788     9921      9918    \
id                                                                             
10094   0.237687  0.303164  0.180006  0.214309  0.186663  0.208990  0.193601   
64032   0.206123  0.185041  0.476707  0.310756  0.282827  0.233870  0.213771   
10167   0.211656  0.172023  0.326409  0.281509  0.329167  0.244354  0.220654   
10031   0.236149  0.189373  0.229253  0.265396  0.286501  0.285474  0.271474   
9888    0.301877  0.243726  0.292043  0.232900  0.362185  0.309011  0.238626   
...          ...       ...       ...       ...       ...       ...       ...   
11359   0.240996  0.190119  0.171304  0.209229  0.183197  0.197432  0.201472   
11360   0.136169  0.132073  0.119405  0.125069  0.121538  0.133738  0.151298   
11364   0.300615  0.241524  0.184869  0.212819  0.193923  0.239664  0.247377 

### Atacantes

In [65]:
# definindo uma variável que exibe um Dataframe com as colunas em interesse
atacantes=jogadores_ok.loc[jogadores_ok["posição"]=="Attacker", ["id", "idade", "chutes/min", "chutes_no_gol/min", "gols/min", "assistências/min", "dribles_tentados/min", "dribles_realizados/min", "faltas_sofridas/min"]]
atacantes.reset_index(drop=True, inplace=True)

In [66]:
# verificando o novo Dataframe
atacantes

id  idade  chutes/min  chutes_no_gol/min  gols/min  assistências/min  \
0     10258     26    0.024557           0.010914  0.004093          0.001364   
1     10041     32    0.032814           0.016129  0.003893          0.000556   
2     10578     28    0.013447           0.006369  0.001415          0.000708   
3     10042     29    0.020266           0.011181  0.002096          0.001398   
4    197059     22    0.036011           0.011080  0.005540          0.000000   
..      ...    ...         ...                ...       ...               ...   
632   11683     21    0.007112           0.002845  0.001422          0.000000   
633    6529     33    0.026949           0.018287  0.008662          0.000000   
634   11369     35    0.012007           0.006289  0.002859          0.002859   
635   35956     35    0.017637           0.007643  0.001176          0.001764   
636   11377     21    0.020927           0.016442  0.005979          0.001495   

     dribles_tentados/min  dribles_realizados/min  faltas_sofridas/min  
0                0.050477                0.031378             0.021828  
1                0.012792                0.007786             0.022803  
2                0.021231                0.012031             0.020524  
3                0.037037                0.027254             0.027952  
4                0.063712                0.038781             0.027701  
..                    ...                     ...                  ...  
632              0.024182                0.007112             0.018492  
633              0.008181                0.003369             0.015399  
634              0.021727                0.012579             0.017724  
635              0.019988                0.014109             0.036449  
636              0.013453                0.005979             0.017937  

[637 rows x 9 columns]

#### Similaridade dos atacantes

In [67]:
# criando uma variável que exibe a normalização de todos os dados menos o id
scaled_atacantes=scaler.fit_transform(atacantes.drop(columns="id"))

In [68]:
# renomeando a variável de atacantes para seus respectivos id e os valores normalizados
atacantes=atacantes[["id"]].merge(pd.DataFrame(scaled_atacantes, columns=atacantes[["idade", "chutes/min", "chutes_no_gol/min", "gols/min", "assistências/min", "dribles_tentados/min", "dribles_realizados/min", "faltas_sofridas/min"]].columns), on=atacantes.index).drop("key_0", axis=1)

In [69]:
# transformando o id em index
atacantes.index=atacantes.id
atacantes.drop(columns="id", inplace=True)

In [70]:
# criando uma variável que exibe um Dataframe que mostra os id similares de cada id
similaridade_atacantes=criar_similaridade(atacantes)

In [71]:
# verificando o novo Dataframe
similaridade_atacantes

id        10258     10041     10578     10042     197059    44492     844     \
id                                                                             
10258   1.000000  0.205820  0.232088  0.392223  0.304151  0.165342  0.183974   
10041   0.205820  1.000000  0.255531  0.238405  0.163720  0.192753  0.292044   
10578   0.232088  0.255531  1.000000  0.298769  0.166746  0.331624  0.343426   
10042   0.392223  0.238405  0.298769  1.000000  0.223046  0.191943  0.218767   
197059  0.304151  0.163720  0.166746  0.223046  1.000000  0.129486  0.138808   
...          ...       ...       ...       ...       ...       ...       ...   
11683   0.191974  0.191076  0.350963  0.213200  0.155133  0.330259  0.245930   
6529    0.166334  0.292072  0.184460  0.171719  0.143996  0.156721  0.218210   
11369   0.207876  0.219378  0.281153  0.253565  0.143722  0.228878  0.288407   
35956   0.211656  0.248492  0.313371  0.296880  0.153069  0.216746  0.263054   
11377   0.204564  0.264780  0.231350  0.210433  0.162067  0.195094  0.249538   

id        10007     10077     159718    10144     9329      10143     1793    \
id                                                                             
10258   0.155525  0.243699  0.222282  0.251495  0.242154  0.353261  0.184104   
10041   0.187612  0.200360  0.271813  0.242038  0.387837  0.253402  0.286222   
10578   0.140540  0.222887  0.359745  0.502549  0.280715  0.234909  0.237729   
10042   0.146012  0.244957  0.263701  0.329475  0.259860  0.325188  0.203187   
197059  0.154624  0.206883  0.173908  0.173000  0.188738  0.292187  0.140681   
...          ...       ...       ...       ...       ...       ...       ...   
11683   0.133170  0.206469  0.302456  0.308565  0.227172  0.196306  0.185984   
6529    0.229982  0.154177  0.201974  0.170497  0.250884  0.196623  0.296308   
11369   0.128086  0.162983  0.224775  0.281820  0.211823  0.183575  0.276293   
35956   0.128850  0.173901  0.268612  0.301770  0.225135  0.207461  0.229314   
11377   0.223810  0.200123  0.281782  0.231699  0.355147  0.225114  0.262210   

id        180235    9936      70314     10359     9585      53958     10098   \
id                                                                             
10258   0.323280  0.220387  0.290319  0.202023  0.176717  0.216091  0.147118   
10041   0.183744  0.326147  0.191109  0.286881  0.314141  0.259591  0.180310   
10578   0.254366  0.279694  0.185846  0.432164  0.270222  0.379210  0.237008   
10042   0.280752  0.254370  0.272945  0.240462  0.202486  0.240348  0.163941   
197059  0.225919  0.160243  0.215956  0.156822  0.140684  0.164963  0.119506   
...          ...       ...       ...       ...       ...       ...       ...   
11683   0.252553  0.197414  0.158049  0.326948  0.203849  0.313184  0.213869   
6529    0.157544  0.257408  0.149289  0.221066  0.248098  0.205019  0.169534   
11369   0.202410  0.328957  0.181825  0.238284  0.230976  0.227606  0.213758   
35956   0.194576  0.292767  0.189452  0.260370  0.228724  0.220908  0.190861   
11377   0.230191  0.241138  0.204467  0.258000  0.207269  0.243904  0.162399   

id        9587      10398     51701     10003     10104     10322     39807   \
id                                                                             
10258   0.161755  0.320259  0.263248  0.329231  0.170474  0.258538  0.196302   
10041   0.230207  0.166991  0.193395  0.248641  0.245245  0.181436  0.193882   
10578   0.259509  0.228771  0.276827  0.268454  0.249412  0.254874  0.367196   
10042   0.194472  0.316029  0.319276  0.318089  0.210155  0.254168  0.227863   
197059  0.133606  0.229949  0.215464  0.252723  0.129167  0.178180  0.159096   
...          ...       ...       ...       ...       ...       ...       ...   
11683   0.212030  0.198791  0.243523  0.232362  0.182033  0.234495  0.502102   
6529    0.191200  0.140197  0.147046  0.195908  0.195162  0.149774  0.151177   
11369   0.214866  0.201192  0.187729  0.189957  0.338435  0.228347  0.197836 

## PostgreSQL

### Ajustes

In [83]:
pip install psycopg2-binary

     |████████████████████████████████| 1.5 MB 4.2 MB/s eta 0:00:01
You should consider upgrading via the '/Library/Frameworks/Python.framework/Versions/3.8/bin/python3 -m pip install --upgrade pip' command.
Note: you may need to restart the kernel to use updated packages.


In [86]:
import sqlalchemy as db

# variáveis necessárias
servidor="postgresql"
usuário="postgres"
senha="admin"#os.getenv("pg_admin")
ip="localhost"
nome="Jogadores América do Sul"

# criando diretório
diretório=db.create_engine(f"{servidor}://{usuário}:{senha}@{ip}/{nome}")

# abrindo a conexão
conn=diretório.connect()

OperationalError: (psycopg2.OperationalError) FATAL:  password authentication failed for user "postgres"

(Background on this error at: http://sqlalche.me/e/13/e3q8)

### Exportando Dataset

In [72]:
# criando um Dataframe com as colunas em interesse a ser exibida no aplicativo
jogadores_ok2=jogadores_ok.loc[:, ['id', 'nome', 'posição', 'idade', 'data_nascimento', 'nacionalidade',
       'time', 'liga', 'total_jogos', 'jogos_escalados', 'minutos_jogados', 'gols_sofridos', 'defesas',
       'gols', 'assistências', 'total_chutes', 'chutes_no_gol', 'dribles_tentados', 'dribles_realizados',
       'total_passes', 'passes_chaves', '%passes_certos', 'desarmes', 'faltas_sofridas', 'faltas_cometidas']]

In [73]:
# exportando os Dataframe
jogadores_ok.to_csv("jogadores.csv", index=False)
jogadores_ok2.to_csv("jogadores2.csv", index=False)
similaridade_goleiros.to_csv("similaridade_goleiros.csv")
similaridade_defesa.to_csv("similaridade_defesa.csv")
similaridade_meias.to_csv("similaridade_meias.csv")
similaridade_atacantes.to_csv("similaridade_atacantes.csv")